src/hotspot/share/opto/loopTransform.cpp

```diff
@@ -3686,6 +3686,14 @@ bool IdealLoopTree::do_remove_empty_loop(PhaseIdealLoop *phase) {
   Node* exact_limit = phase->exact_limit(this);
   Node* final_iv = new SubINode(exact_limit, cl->stride());
   phase->register_new_node(final_iv, cl->in(LoopNode::EntryControl));
+
+  Node* castii = phi->in(LoopNode::EntryControl);
+  // if phi has CastII with carry_dependency, create a new CastII to pin final_iv
+  if (castii->is_CastII() && castii->as_CastII()->carry_dependency()) {
+    Node* cast = ConstraintCastNode::make(castii->in(0), exact_limit, phase->_igvn.type(exact_limit), ConstraintCastNode::UnconditionalDependency, T_INT);
+    phase->_igvn.register_new_node_with_optimizer(cast);
```

Comment:
We should use `register_new_node` here to also set ctrl.
```suggestion
    phase->register_new_node(cast, castii->in(0));
```